src/flp/types/fixedpoint_l2.rs

```diff
@@ -0,0 +1,638 @@
+//! 1. Clients have a vector whose entries are fixed point numbers. Only those
+//!    fixed point types are supported where the numbers lie in the range
+//!    `[-1,1)`.
+//! 2. Because norm computation happens in the validation circuit, it is done
+//!    on entries encoded as field elements. That is, the same vector entries
+//!    are now represented by integers in the range `[0,2^n)`, where `-1` is
+//!    represented by `0` and `+1` by `2^n`.
+//! 3. Because the field is not necessarily exactly of size `2^n`, but might be
+//!    larger, it is not enough to encode a vector entry as in 2. and submit
+//!    it to the aggregator. Instead, in order to make sure that all submitted
+//!    values are in the correct range, they are bit-encoded. (This is the same
+//!    as what happens in the `Sum` type.)
+//!    This means that instead of sending a field element in the range `[0,2^n)`,
+//!    we send `n` field elements representing the bit encoding. The validation
+//!    circuit can verify that all submitted "bits" are indeed either `0` or `1`.
+//! 4. The computed and submitted norms are treated very similar to the vector
+//!    entries, but they have a different number of bits, namely `2n-2`.
+//! 5. As the aggregation result is a pointwise sum of the client vectors,
+//!    the numbers no longer (semantically) lie in the range `[-1,1)`, and cannot
+//!    be represented by the same fixed point type as the input. Instead the
+//!    decoding happens directly into a vector of floats.
+//!
+//! ### Fixed point encoding
+//! Submissions consist of encoded fixed-point numbers in `[-1,1)` represented as
+//! field elements in `[0, 2^n)`, where n is the number of bits the fixed-point
+//! representation has. Encoding and decoding is handled by the associated functions of the
+//! [`CompatibleFloat`](crate::flp::types::fixedpoint_l2::compatible_float::CompatibleFloat)
+//! trait. Semantically, the following function describes how a fixed-point value `x`
+//! in range `[-1,1)` is converted to a field integer:
+//! ```text
+//! enc : [-1,1) -> [0,2^n)
+//! enc(x) = 2^(n-1) * x + 2^(n-1)
+//! ```
+//! The inverse is:
+//! ```text
+//! dec : [0,2^n) -> [-1,1)
+//! dec(y) = (y - 2^(n-1)) * 2^(1-n)
+//! ```
+//! Note that these functions only make sense when interpreting all occuring
+//! numbers as real numbers. Since our signed fixed-point numbers are encoded as
+//! two's complement integers, the computation that happens in
+//! [`CompatibleFloat::to_field_integer`](crate::flp::types::fixedpoint_l2::compatible_float::CompatibleFloat::to_field_integer)
+//! is actually simpler.
+//!
+//! ### Norm computation
+//! The L2 norm of a vector xs of numbers in `[-1,1)` is given by:
+//! ```text
+//! norm(xs) = sqrt(sum_{x in xs} x^2)
+//! ```
+//! Instead of computing the norm, we make two simplifications:
+//! 1. We ignore the square root, which means that we are actually computing
+//!    the square of the norm.
```

Comment:
or square the bound. but since we're only comparing to the client submission who also computes the square, it'd be fine. other bounds would require some extra work tho as the bound 1 is actually only enforced by the encoding size